tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/Reduce.java

```diff
@@ -0,0 +1,239 @@
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the type for the variables and result
+   */
+  protected Reduce(Ops tf, String name, MetricReduction reduction, long seed, Class<T> type) {
+    super(tf, name, seed);
+    this.reduction = reduction;
+    this.totalName = this.getVariableName(TOTAL);
+    this.countName = this.getVariableName(COUNT);
+    this.type = type;
+    setupVars();
+  }
+  /** Initializes the Variables */
+  private void setupVars() {
+    if (total == null) {
+      total = getTF().withName(totalName).variable(Shape.scalar(), type);
+    }
+    if (reduction == MetricReduction.SUM_OVER_BATCH_SIZE
+        || reduction == MetricReduction.WEIGHTED_MEAN) {
+      if (count == null) {
+        count = getTF().withName(countName).variable(Shape.scalar(), type);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public Op resetStates() {
+    List<Op> controls = new ArrayList<>();
+    if (total != null) {
+      controls.add(
+          getTF().assign(total, CastHelper.cast(getTF(), getTF().constant(0), total.type())));
+    }
+    if (count != null) {
+      controls.add(
+          getTF().assign(count, CastHelper.cast(getTF(), getTF().constant(0), count.type())));
+    }
+    return getTF().withControlDependencies(controls).noOp();
+  }
+
+  /**
+   * Updates the metric variables based on the inputs. At least one input arg required for <code>
+   * values</code>, an optional additional input for the <code>sampleWeights</code>
+   *
+   * @param values the inputs to be passed to update state, this may not be null
+   * @param sampleWeights sample weights to be applied to values, may be null.
+   * @return the result with a control dependency on update state Operands
+   * @throws IllegalArgumentException if values is null
+   */
+  @Override
+  public List<Op> updateStateList(Operand<U> values, Operand<T> sampleWeights) {
+
+    if (values == null) {
+      throw new IllegalArgumentException("values is required.");
+    }
+    List<Op> updateOperations = new ArrayList<>();
+    // cast everything to match the variables
+    Operand<U> lSampleWeights = null;
+    Operand<U> lValues = values;
+
+    if (sampleWeights != null) {
+      lSampleWeights = CastHelper.cast(getTF(), sampleWeights, lValues.type());
+      LossTuple<U> tuple =
+          LossesHelper.squeezeOrExpandDimensions(getTF(), null, lValues, lSampleWeights);
+      lValues = tuple.getTarget();
+      lSampleWeights = tuple.getSampleWeights();
+      try {
+
+        Op broadcastWeightsCheck = MetricsHelper.broadcastWeights(getTF(), lSampleWeights, lValues);
+        lValues =
+            getTF()
+                .withSubScope("broadcastWeightsCheck")
+                .withControlDependencies(Collections.singletonList(broadcastWeightsCheck))
+                .math
+                .mul(lValues, lSampleWeights);
+      } catch (IllegalArgumentException ex) {
```

Comment:
It feels unsafe to me to assume the only possible reason for an `IllegalArgumentException` in this try/catch is the one handled here. Instead, I would think we'd want to explicitly check for this condition?

Reply:
I understand your point. Should we create something like `BroadcastException`?

Reply:
A special-purpose exception would certainly make this safe and more explicit, and I'd be fine with it. To me, encapsulating this additional information in the return type would feel more idiomatic, but for this intendedly-module-private method, I don't see that as a strong argument.

Reply:
There is already `TFInvalidArgumentException` which extends `IllegalArgumentException` that is thrown for the dynamic analysis part of this function thrown from within the graph. Should we throw `TFInvalidArgumentException` instead for static shapes? Though, that is still somewhat vague, so maybe `BroadcastException` makes more sense.

Reply:
Yeah, I'd be a fan of the more specific exception name. Perhaps even `NotBroadcastableException`?

Reply:
Added `NotBroadcastableException` which is thrown if the static shapes are not broadcastable.  `TFInvalidArgumentException` is still thrown when evaluating dynamic shapes as that is done inside the graph. Both exceptions inherit from `IllegalArgumentException`.

Reply:
ok